File: python/Cell.py
from typing import Optional

from Distances import Distances


class Cell:
    """
    Cell represents a cell in a two-dimensional Grid. Every cell knows its row and column position in the Grid,
    as well as if it has neighboring cells to the North, South, East or West. Cells also keep track of any "links"
    to those neighboring Cells. A link means that a "passage" has been "carved" between this cell and another cell,
    connecting them in a maze.
    Two Cells are considered equal if their corresponding row,col indices are equal.
    """

    def __init__(self, row: int, col: int):
        self.row = row
        self.col = col
        self.north: Optional[Cell] = None
        self.south: Optional[Cell] = None
        self.east: Optional[Cell] = None
        self.west: Optional[Cell] = None
        self.links = {}

    def link(self, other, bidi=True):
        """
         creates a link between this Cell and other. Use this method to "carve a passage" between cells
        :param other: the other Cell to carve a passage to
        :param bidi: bidirectional, if True, than also carve a passage between other and self
        :return:
        """
        self.links[other] = True
        if bidi:
            other.links[self] = True

    def unlink(self, other, bidi=True):
        """
        removes a link between this cell and other. If the link did not exist to begin with then Raises a KeyError
        :param other: the other Cell to unlink from
        :param bidi: bidirectional, if True, then also remove the link from other to self
        :return:
        """
        del self.links[other]
        if bidi:
            del other.links[self]

    def linked_cells(self):
        """
        :return: a list of all keys currently in this Cell's links dictionary
        """
        return list(self.links)

    def is_linked(self, other):
        """
        checks if this Cell has a link to other
        :param other: the other Cell to test
        :return: True if there is a link from this Cell to other, else False
        """
        return other in self.links

    def neighbors(self):
        """
        :return: a list containing all neighbor cells of this cell. Neighbors that do not exist (i.e. None)
        are not returned
        """
        neighbors = []
        if self.north is not None:
            neighbors.append(self.north)
        if self.south is not None:
            neighbors.append(self.south)
        if self.east is not None:
            neighbors.append(self.east)
        if self.west is not None:
            neighbors.append(self.west)
        return neighbors

    def distances(self) -> Distances:
        """
        computes the distances between this cell and every other cell that is linked to this cell
        :return: a Distances object, with this cell as the root, and with distances computed to every other linked cell
        """
        distances = Distances(self)
        frontier = [self]

        while frontier:
            new_frontier = []

            for cell in frontier:
                for linked_cell in cell.linked_cells():
                    if linked_cell not in distances.cells:
                        distances.cells[linked_cell] = distances.cells[cell] + 1
                        new_frontier.append(linked_cell)

            frontier.clear()
            frontier.extend(new_frontier)

        return distances

    def __str__(self):
        return "({},{}) N:{} S:{} E:{} W:{} links:{}".format(
            self.row,
            self.col,
            self.north is not None,
            self.south is not None,
            self.east is not None,
            self.west is not None,
            len(self.links)
        )

    def __eq__(self, other: object) -> bool:
        """ two cells are equal of their respective row and column values are equal"""
        if not isinstance(other, Cell):
            return NotImplemented
        return self.row == other.row and self.col == other.col

    def __ne__(self, other: object) -> bool:
        if not isinstance(other, Cell):
            return NotImplemented
        return self.row != other.row or self.col != other.col

    def __hash__(self) -> int:
        return self.row.__hash__() + self.col.__hash__()

File: nodejs/src/Cell.js
const Distances = require("./Distances");

/**
 * Cell is the main object used in building mazes.
 * Cell's know their row and column position in a Grid. They know if they have a neighboring cell (or not) to the
 * North, South, East, and West. Most importantly, Cells keep track of any "links" to those neighbors.
 * A link means that a passageway has been carved between this cell and a neighboring cell. Once multiple Cells are
 * linked together they will form a maze
 *
 */
class Cell {

    /**
     * constructs a new Cell at the specified row and column index. All neighbors are set to null, the cell's links
     * are empty and the Cell's weight is set to 1.
     * @param row {Integer} this Cell's row index in a Grid
     * @param col {Integer} this Cell's column index in a Grid
     */
    constructor(row, col) {
        this._row = row;
        this._col = col;
        this._links = new Map();
        this._north = null;
        this._south = null;
        this._east = null;
        this._west = null;
        this._weight = 1;
    }


    get row () {
        return this._row;
    }

    get col () {
        return this._col;
    }

    /**
     * create a link between this cell and other.
     * @param other {Cell} the other Cell you want to create a link to
     * @param bidi {boolean} bi-directional link, if true, then also create a link between other and this Cell
     */
    link(other, bidi=true) {
        this._links.set(other, true);
        if (bidi) {
            other.link(this, false);
        }
    }

    /**
     * removes a link between this cell and other. If the link did not exist to begin with, then nothing happens
     * @param other {Cell} the other Cell to unlink from
     * @param bidi {boolean} bi-directonal link, if true, then this method will also attempt to remove the link between other
     *             and this cell
     */
    unlink(other, bidi=true) {
        this._links.delete(other);
        if (bidi) {
            other.unlink(this, false);
        }
    }

    /**
     * returns an Array containing all the cells that this cell links to
     */
    links() {
        return [...this._links.keys()];
    }

    /**
     * is this Cell linked to other cell?
     * @param otherCell
     * @returns {boolean} true if this Cell is linked to otherCell
     */
    isLinked(otherCell) {
        return this._links.has(otherCell);
    }

    /**
     * @returns {*[]} an array containing the Cells that are neighbors of this Cell. Neighbors that are null,
     * are NOT returned
     */
    neighbors() {
        let neighbors = [this._north, this._south, this._east, this._west];
        return neighbors.filter(neighbor => neighbor)
    }

    /**
     * computes the distances (i.e. the cost of moving into linked cells), using this cell
     * as the root cell.
     * @returns {Distances} a Distances object, with this cell as the root, and with distances (weights) computed for
     * every other cell in the grid.
     */
    distances() {
        // weights holds the current weights for each cell in the grid
        const weights = new Distances(this);

        // pending holds the cells of the grid that need to be visited
        const pending = [this];

        while (pending.length > 0) {
            // sort by cell weight, descending order
            let current = pending.sort((cell1, cell2) => cell2.weight - cell1.weight).pop();

            // iterate through the linked neighbors of current and find the neighbor with the lowest cost
            for(const neighbor of current.links()) {
                // the total weight of moving into a neighboring cell is the total weight
                // of the current path so far, plus the weight of the neighbor
                const totalWeight = weights.distanceOf(current) + neighbor.weight;

                if (!weights.contains(neighbor) || totalWeight < weights.distanceOf(neighbor)) {
                    pending.push(neighbor);
                    weights.put(neighbor, totalWeight);
                }
            }
        }
        return weights;
    }

    get north() {
        return this._north;
    }

    /**
     * sets this Cell's north neighbor to the passed in cell
     * @param cell {Cell} a cell to set as this cell's neighbor. If this cell does not have a neighbor then set
     * this to null
     */
    set north(cell) {
        this._north = cell;
    }

    get south() {
        return this._south;
    }

    /**
     * sets this Cell's south neighbor to the passed in cell
     * @param cell {Cell} a cell to set as this cell's neighbor. If this cell does not have a neighbor then set
     * this to null
     */
    set south(cell) {
        this._south = cell;
    }

    get east() {
        return this._east;
    }

    /**
     * sets this Cell's east neighbor to the passed in cell
     * @param cell {Cell} a cell to set as this cell's neighbor. If this cell does not have a neighbor then set
     * this to null
     */
    set east(cell) {
        this._east = cell;
    }

    get west() {
        return this._west;
    }

    /**
     * sets this Cell's west neighbor to the passed in cell
     * @param cell {Cell} a cell to set as this cell's neighbor. If this cell does not have a neighbor then set
     * this to null
     */
    set west(cell) {
        this._west = cell;
    }

    get weight() {
        return this._weight;
    }

    /**
     * set this Cell's weight to value
     * @param value {number}
     */
    set weight(value) {
        this._weight = value;
    }

    toString() {
        return `Cell{ +
            "row=${this._row}
            ", col=${this._col}
            ", linkCount=${this._links.size}
            '}`;
    }

    // /**
    //  * two Cells are equal if their respective row,column values are equal
    //  * @param cell {Cell}
    //  * @returns {boolean}
    //  */
    // equals(cell) {
    //     return this._row === cell.row && this._col === cell.col;
    // }
}

module.exports = Cell;